core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1337,290 +1451,127 @@ class Block : public lifecycle::StateMachine<Derived>, //
+                constexpr std::size_t                          max_simd_double_size  = stdx::simd_abi::max_fixed_size<double>;
+                constexpr std::size_t                          simd_size             = input_types_simd_size == 0 ? max_simd_double_size : std::min(max_simd_double_size, input_types_simd_size * 4);
+                std::integral_constant<std::size_t, simd_size> width{};
+
+                if constexpr ((meta::simdize_size_v<output_simd_types> != 0) and ((requires(Derived &d) {
+                    { d.processOne_simd(simd_size) };
+                }) or (meta::simdize_size_v<input_simd_types> != 0 and traits::block::can_processOne_simd<Derived>))) { // SIMD loop
+                    ret = invokeProcessOneSimd(inputSpans, outputSpans, width, processedIn);
+                } else {                                                 // Non-SIMD loop
+                    if constexpr (HasConstProcessOneFunction<Derived>) { // processOne is const -> can process whole batch similar to SIMD-ised call
+                        ret = invokeProcessOnePure(inputSpans, outputSpans);
+                    } else { // processOne isn't const i.e. not a pure function w/o side effects -> need to evaluate state after each sample
+                        const auto result = invokeProcessOneNonConst(inputSpans, outputSpans, processedIn);
+                        ret = result.status;
+                        processedIn = result.processedIn;
+                        processedOut = result.processedOut;
                     }
                 }
             }
-
-            //            if constexpr (kIsSourceBlock) {
-            if (nOutSamplesBeforeRequestedStop > 0) {
-                ports_status.out_samples = nOutSamplesBeforeRequestedStop;
-                nSamplesToConsume        = nOutSamplesBeforeRequestedStop;
-            }
-            //            }
-
-            forwardTags();
-            write_to_outputs(ports_status.out_samples, writersTuple);
-            const bool success = consumeReaders(self(), nSamplesToConsume);
-            if (lifecycle::isShuttingDown(this->state())) [[unlikely]] {
-                if (auto e = this->changeStateTo(lifecycle::State::STOPPED); !e) {
-                    emitErrorMessage("isShuttingDown -> STOPPED", e.error());
-                }
-                publishTag({ { gr::tag::END_OF_STREAM, true } }, 0);
-                return { requested_work, ports_status.in_samples, success ? work::Status::DONE : work::Status::ERROR };
+        } else { // block does not define any valid processing function
+            static_assert(gr::meta::always_false<gr::traits::block::stream_input_port_types_tuple<Derived>>, "neither processBulk(...) nor processOne(...) implemented");
+        }
+        forwardTags();
+        if (lifecycle::isShuttingDown(this->state()) || nextEosTag <= processedIn + 1) {
```

Comment:
Here something is going wrong for my Delay block (processBulk(ConsumableSpan, PublishableSpan): The block produces/consumes 0, but both `processedIn` and `processedOut` are `100`, which is the data available at the input port. This check then considers the EOS tag in range and finishes.

Reply:
Yes, see also this comment: https://github.com/fair-acc/graph-prototype/pull/297/files#diff-d0a6abe85a332778390dc74e1c498e4c580677fe46db76395cdf2afefc41fb1eR1507 Until now, this has not been a problem, because there was no block that relied on being able to not consume all input samples in process bulk. Probably the easiest and most generic way to fix it would be to remove the eosTag condition and only shutdown the block in the next call to process bulk...  have to check if this breaks something else.